### ISSUE TITLE
Correct Lifecycle flow at bootstrap time

### ISSUE THREAD

**Issue by matthias-reis** [xing/hops#260]
## This is my intent (choose one)

- [x] I want to report a bug
- [ ] I want to request a feature or change
- [ ] I want to provide or change a feature

## The problem

Before Hops V 7, there was a clear order for the calls of lifecycle methods. The async `bootstrap` had been called _and finished_ before `enhanceElement`. This is currently used to fetch async data before rendering to the DOM, so Hops 7 it breaks in current applications.

## Proposed solution

Three scenarios should be supported:

- doing asynchronous stuff __before creating__  the element (see example above)
- doing asynchronous staff __before rendering__ and __after creating__ the element (usecase: apollo graphql client)
- composition must still be supported

__Possible solution:__
Turn `enhanceElement` into a Promise, that receives a Promise from the chain before and returns again a Promise. This Promise should in any case resolve to a renderable Element.

__Example (no runnable code, just to explain the principle):__

```jsx
// using jsx for clarification, should be es5
enhanceElement: function (prev) {
  let data = {};

  return fetchSth()
    .then(remoteData => data = remoteData)
    .then(prev)
    .then(el => <MyProvider data={data}>{el}</MyProvider>)
    .then(el => {
      return doSthAsync()
        .then(Promise.resolve(el));
    });
}
```

**Comment by dmbch** [comment]
Hey @matthias-reis, thanks for reporting this! I really like your idea, but I thought this was already fixed in https://github.com/xing/hops/pull/254 and released with hops-react v7.3.2. Can you confirm?

**Comment by matthias-reis** [comment]
Yepp, you"re right ...
